### PR TITLE
Feat/dashboard supports random port

### PR DESCRIPTION
**Pull request by imneov**
# Description

1. support `dashboardLocalPort` set to `0`
2. update `pf.LocalPort` when `PortForward` is `Ready`

## Issue reference

Please reference the issue this PR will close: #863 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Comment by mukundansundar** [comment]
@imneov Please fix linter errors ...

File: pkg/kubernetes/portforward.go [hunk 1]
```diff
@@ -114,27 +115,29 @@ func (pf *PortForward) run() error {
...
 
 	select {
 	// if `pf.run()` succeeds, block until terminated
 	case <-pf.ReadyCh:
+		ports, err := fw.GetPorts()
+		if err != nil {
+			return fmt.Errorf("can't get the ports that were forwarded: %w", err)
+		}
+		if len(ports) == 0 {
+			return fmt.Errorf("error ports length")
```
Comment (mukundansundar) [inline pkg/kubernetes/portforward.go:136]:
```suggestion
			return fmt.Errorf("can not get the ports that were forwarded: error getting ports length")
```

File: pkg/kubernetes/portforward.go [hunk 2]
```diff
@@ -114,27 +115,29 @@ func (pf *PortForward) run() error {
...
-
 	go func() {
-		if err := pf.run(); err != nil {
+		if err := fw.ForwardPorts(); err != nil {
 			failure <- err
 		}
 	}()
 
 	select {
 	// if `pf.run()` succeeds, block until terminated
```
Comment (mukundansundar) [inline pkg/kubernetes/portforward.go:129]:
```suggestion
	// if `fw.ForwardPorts()` succeeds, block until terminated
```

File: pkg/kubernetes/portforward.go [hunk 3]
```diff
@@ -114,27 +115,29 @@ func (pf *PortForward) run() error {
 
 	fw, err := portforward.NewOnAddresses(dialer, []string{pf.Host}, ports, pf.StopCh, pf.ReadyCh, out, errOut)
 	if err != nil {
-		return err
+		return fmt.Errorf("can't creat PortForwarder: %w", err)
```
Comment (mukundansundar) [inline pkg/kubernetes/portforward.go:118]:
```suggestion
		return fmt.Errorf("can not create PortForwarder: %w", err)
```

File: pkg/kubernetes/portforward.go [hunk 4]
```diff
@@ -94,12 +94,13 @@ func NewPortForward(
...
-// until Stop() is called.
-func (pf *PortForward) run() error {
+// Init creates and runs a port-forward connection.
+// This function blocks until connection is established.
+// Note: Caller should call Stop() to finish the connection.
+func (pf *PortForward) Init() error {
 	transport, upgrader, err := spdy.RoundTripperFor(pf.Config)
 	if err != nil {
-		return err
+		return fmt.Errorf("can't connect to a Kubernetes cluster: %v", err)
```
Comment (mukundansundar) [inline pkg/kubernetes/portforward.go:103]:
```suggestion
		return fmt.Errorf("can not connect to Kubernetes cluster: %w", err)
```

File: pkg/kubernetes/portforward.go [hunk 5]
```diff
@@ -52,7 +52,7 @@ func NewPortForward(
 ) (*PortForward, error) {
 	client, err := k8s.NewForConfig(config)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("can't creat Clientset: %s", deployName)
```
Comment (mukundansundar) [inline pkg/kubernetes/portforward.go:55]:
```suggestion
		return nil, fmt.Errorf("can't create Clientset for %q", deployName)
```

**Review by mukundansundar — COMMENTED** [review]
@imneov Thanks for the contribution.
Overall looks good. Just some small changes.

**Comment by imneov** [comment]
@mukundansundar Thanks for the suggestions.

**Comment by mukundansundar** [comment]
@imneov Please do fix the DCO and the conflicts ...

**Comment by imneov** [comment]
> @imneov Please do fix the DCO and the conflicts ...

OK, It`s done.

File: pkg/kubernetes/portforward.go [hunk 6]
```diff
@@ -114,27 +115,29 @@ func (pf *PortForward) run() error {
...
 		}
 	}()
 
 	select {
-	// if `pf.run()` succeeds, block until terminated.
+	// if `fw.ForwardPorts()` succeeds, block until terminated
 	case <-pf.ReadyCh:
+		ports, err := fw.GetPorts()
+		if err != nil {
+			return fmt.Errorf("can't get the ports that were forwarded: %w", err)
```
Comment (mukundansundar) [inline pkg/kubernetes/portforward.go:133]:
```suggestion
			return fmt.Errorf("can not get the local and remote ports: %w", err)
```

File: pkg/kubernetes/portforward.go [hunk 7]
```diff
@@ -114,27 +115,29 @@ func (pf *PortForward) run() error {
...
+		}
+		if len(ports) == 0 {
+			return fmt.Errorf("can not get the ports that were forwarded: error getting ports length")
```
Comment (mukundansundar) [inline pkg/kubernetes/portforward.go:136]:
```suggestion
			return fmt.Errorf("can not get the local and remote ports: error getting ports length")
```

**Review by mukundansundar — COMMENTED** [review]
@imneov Almost done ... just a few small changes and linter fixes ...

**Comment by imneov** [comment]
> @imneov Almost done ... just a few small changes and linter fixes ...

@mukundansundar Thank you kindly!   I fix it.

**Comment by mukundansundar** [comment]
> > @imneov Almost done ... just a few small changes and linter fixes ...
> 
> @mukundansundar Thank you kindly! I fix it.

@imneov Let me know if you will be able to address some of the required changes ... Thanks

**Comment by pravinpushkar** [comment]
LGTM
However, There will be a need to update the documentation here - https://docs.dapr.io/reference/cli/dapr-dashboard/ .
@imneov Would you be able to create a doc issue here - https://github.com/dapr/docs/issues and a small PR to update the same.

File: cmd/dashboard.go [hunk 8]
```diff
@@ -84,7 +87,7 @@ dapr dashboard -k -p 9999
 			os.Exit(1)
 		}
 
-		if dashboardLocalPort <= 0 {
+		if dashboardLocalPort < 0 {
```
Comment (shubham1172) [inline cmd/dashboard.go:90]: How does this work for standalone mode? Standalone mode calls the dashboard CLI directly and passes a zero value to this code:
https://github.com/dapr/dashboard/blob/febc4a7bca4cb992bc3fdd9799a0a87b5922f8f6/cmd/webserver.go#L72

Have you tried running it locally?

File: pkg/kubernetes/portforward.go [hunk 9]
```diff
@@ -52,7 +52,7 @@ func NewPortForward(
 ) (*PortForward, error) {
 	client, err := k8s.NewForConfig(config)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("can't create Clientset for %q", deployName)
```
Comment (shubham1172) [inline pkg/kubernetes/portforward.go:55]: We are losing the `err` value here.

File: pkg/kubernetes/portforward.go [hunk 10]
```diff
@@ -94,12 +94,13 @@ func NewPortForward(
...
-// until Stop() is called.
-func (pf *PortForward) run() error {
+// Init creates and runs a port-forward connection.
+// This function blocks until connection is established.
+// Note: Caller should call Stop() to finish the connection.
+func (pf *PortForward) Init() error {
 	transport, upgrader, err := spdy.RoundTripperFor(pf.Config)
 	if err != nil {
-		return err
+		return fmt.Errorf("can not connect to Kubernetes cluster: %w", err)
```
Comment (shubham1172) [inline pkg/kubernetes/portforward.go:103]:
```suggestion
		return fmt.Errorf("cannot connect to Kubernetes cluster: %w", err)
```

File: pkg/kubernetes/portforward.go [hunk 11]
```diff
@@ -114,27 +115,29 @@ func (pf *PortForward) run() error {
 
 	fw, err := portforward.NewOnAddresses(dialer, []string{pf.Host}, ports, pf.StopCh, pf.ReadyCh, out, errOut)
 	if err != nil {
-		return err
+		return fmt.Errorf("can not create PortForwarder: %w", err)
```
Comment (shubham1172) [inline pkg/kubernetes/portforward.go:118]:
```suggestion
		return fmt.Errorf("cannot create PortForwarder: %w", err)
```

**Review by shubham1172 — CHANGES_REQUESTED** [review]
Please address the comments.

File: cmd/dashboard.go (hunk 8 above)
Comment (shubham1172) [reply cmd/dashboard.go:90]: @imneov

**Comment by imneov** [comment]
> Please address the comments.

It's fixed

File: cmd/dashboard.go (hunk 8 above)
Comment (imneov) [reply cmd/dashboard.go:90]: The standalone mode is different. I think there are two main reasons:
1. The dashboard is already running in the K8S environment, so `dashboardLocalPort==0` will not generate a new process. But in standalone mode `dashboardLocalPort==0` will generate a new process,
2. The dashboard already running in standalone mode has no suitable method to obtain the listening port information

Comment (shubham1172) [reply cmd/dashboard.go:90]: When `dashboardLocalPort` is 0 for standalone mode, it passes the port number as is and ends up calling [this](https://github.com/dapr/dashboard/blob/febc4a7bca4cb992bc3fdd9799a0a87b5922f8f6/cmd/webserver.go#L108-L115):

```go
	srv := &http.Server{
		Handler:      r,
		Addr:         fmt.Sprintf("0.0.0.0:%v", port),
		WriteTimeout: 15 * time.Second,
		ReadTimeout:  15 * time.Second,
	}
	fmt.Printf("Dapr Dashboard running on http://localhost:%v\n", port)
	log.Fatal(srv.ListenAndServe())
```
Port 0 should identify the next available free port and start the server. However, the user won't know which port is it running on. The `fmt.Printf` will still print `localhost:0`. We will need to fix this. Can you please create an issue to track the same?

We will need to use `net.Listen` like in https://stackoverflow.com/a/43425461/4654175

cc @mukundansundar

Comment (imneov) [reply cmd/dashboard.go:90]: OK, I will try to add it

Comment (imneov) [reply cmd/dashboard.go:90]: @shubham1172 
I commit changes on a pull request  https://github.com/dapr/dashboard/pull/205/files

**Review by shubham1172 — DISMISSED** [review]

File: cmd/dashboard.go [hunk 12]
```diff
@@ -160,8 +163,8 @@ dapr dashboard -k -p 9999
 				portForward.Stop()
 			}()
 
-			// url for dashboard after port forwarding.
-			var webURL string = fmt.Sprintf("http://%s:%d", dashboardHost, dashboardLocalPort)
+			// url for dashboard after port forwarding
```
Comment (yaron2) [inline cmd/dashboard.go:166]: The linter is failing because this comment doesn't have a dot `.` at the end.

**Review by yaron2 — DISMISSED** [review]

File: cmd/dashboard.go (hunk 12 above)
Comment (imneov) [reply cmd/dashboard.go]: Fix it.

**Comment by pravinpushkar** [comment]
@imneov Please update the branch and resolve the conflicts. Additionally, Please update the documentation also.

**Comment by imneov** [comment]
> @imneov Please update the branch and resolve the conflicts. Additionally, Please update the documentation also.


@pravinpushkar It's done.

**Comment by imneov** [comment]
@pravinpushkar HI~  Is there anything else I need to do?

**Comment by pravinpushkar** [comment]
> @pravinpushkar HI~ Is there anything else I need to do?

There are few PR checks failing, could you please have a look? For now I have re-triggered some of them. Hopefully E2E- Upgrade path tests should pass as it looked temporary failure. But other one should not have failed.

**Comment by imneov** [comment]
> > @pravinpushkar HI~ Is there anything else I need to do?
> 
> There are few PR checks failing, could you please have a look? For now I have re-triggered some of them. Hopefully E2E- Upgrade path tests should pass as it looked temporary failure. But other one should not have failed.

HI, @pravinpushkar ~ I don't find the failing PR checks,could you give me a link?

**Review by mukundansundar — APPROVED** [review]

**Pull request merged**